sql/core/src/main/scala/org/apache/spark/sql/execution/joins/ShuffledHashJoinExec.scala

```diff
@@ -332,6 +327,266 @@ case class ShuffledHashJoinExec(
     HashedRelationInfo(relationTerm, keyIsUnique = false, isEmpty = false)
   }
 
+  override def doProduce(ctx: CodegenContext): String = {
+    // Specialize `doProduce` code for full outer join, because full outer join needs to
+    // iterate streamed and build side separately.
+    if (joinType != FullOuter) {
+      return super.doProduce(ctx)
+    }
+
+    val HashedRelationInfo(relationTerm, _, _) = prepareRelation(ctx)
+
+    // Inline mutable state since not many join operations in a task
+    val keyIsUnique = ctx.addMutableState("boolean", "keyIsUnique",
+      v => s"$v = $relationTerm.keyIsUnique();", forceInline = true)
+    val streamedInput = ctx.addMutableState("scala.collection.Iterator", "streamedInput",
+      v => s"$v = inputs[0];", forceInline = true)
+    val buildInput = ctx.addMutableState("scala.collection.Iterator", "buildInput",
+      v => s"$v = $relationTerm.valuesWithKeyIndex();", forceInline = true)
+    val streamedRow = ctx.addMutableState("InternalRow", "streamedRow", forceInline = true)
+    val buildRow = ctx.addMutableState("InternalRow", "buildRow", forceInline = true)
+
+    // Generate variables and related code from streamed side
+    val streamedVars = genOneSideJoinVars(ctx, streamedRow, streamedPlan, setDefaultValue = false)
+    val streamedKeyVariables = evaluateRequiredVariables(streamedOutput, streamedVars,
+      AttributeSet.fromAttributeSets(HashJoin.rewriteKeyExpr(streamedKeys).map(_.references)))
+    ctx.currentVars = streamedVars
+    val streamedKeyExprCode = GenerateUnsafeProjection.createCode(ctx, streamedBoundKeys)
+    val streamedKeyEv =
+      s"""
+         |$streamedKeyVariables
+         |${streamedKeyExprCode.code}
+       """.stripMargin
+    val streamedKeyAnyNull = s"${streamedKeyExprCode.value}.anyNull()"
```

Comment:
why does `keyIsUnique` need a mutable state but `anyNull` does not?

Reply:
I want to avoid per-row call of `UnsafeHashedRelation.keyIsUnique()`. So store the function call result into a variable of `keyIsUnique`, and only call once for all rows (example - [line 079 of `shj_keyIsUnique_0 `](https://gist.github.com/c21/828b782ee81827f4148939cb50314a7b)). Though `UnsafeHashedRelation.keyIsUnique()` call is very cheap now (just `binaryMap.numKeys() == binaryMap.numValues()`), but it can prevent future bug if the call becomes expensive if we change the implementation later, and it's just looking better.

Reply:
isn't `anyNull` called more frequently than `keyIsUnique`. We call `anyNull` in the while loop;.

Reply:
@cloud-fan - sorry if I am missing anything, `anyNull` returns whether any null of the current input row (`UnsafeRow.anyNull()`). The result is varied for each input row in while loop.

Reply:
ah I misread the code, nvm

Reply:
@cloud-fan - no worries, thanks for the careful review.